File: tools/pack/archive.py
"""Fail-closed `.atlaspack` ZIP preflight and extraction.

Archive bytes are untrusted. This module never executes archive content and never
uses ZipFile.extract()/extractall(): all paths and entry types are validated before
manual streaming into a private staging directory.
"""
from __future__ import annotations

import os
import re
import stat
import unicodedata
import zipfile
from dataclasses import dataclass
from pathlib import Path, PurePosixPath

from tools.pack.contract import ACTIVE_EXTENSIONS, DRIVE_QUALIFIED, WINDOWS_RESERVED

ROOT_NAMESPACES = {"metadata", "targets"}
TARGET_NAMESPACES = {"atlas", "content", "search"}
TOP_LEVEL_METADATA = {"root.json", "timestamp.json", "snapshot.json", "targets.json"}
VERSIONED_ROOT_RE = re.compile(r"^[1-9][0-9]*\.root\.json$")
ALLOWED_COMPRESSION = {zipfile.ZIP_STORED, zipfile.ZIP_DEFLATED}
WINDOWS_REPARSE_POINT = 0x0400
COPY_CHUNK = 1024 * 1024


class PackArchiveError(ValueError):
    """Raised when an `.atlaspack` violates the strict archive safety profile."""


@dataclass(frozen=True)
class PackSafetyLimits:
    """Reference v1 extraction envelope; configurable without changing canonical schemas."""

    profile_version: str = "1.0.0"
    max_archive_bytes: int = 2 * 1024 * 1024 * 1024
    max_entries: int = 8192
    max_path_chars: int = 512
    max_file_uncompressed_bytes: int = 2 * 1024 * 1024 * 1024
    max_total_uncompressed_bytes: int = 4 * 1024 * 1024 * 1024
    max_compression_ratio: float = 200.0


@dataclass(frozen=True)
class ExtractionReport:
    profile_version: str
    file_count: int
    directory_count: int
    total_uncompressed_bytes: int


def _component_safe(part: str) -> None:
    if not part or part in {".", ".."}:
        raise PackArchiveError(f"non-canonical path component: {part!r}")
    stripped = part.rstrip(" .")
    if stripped != part or not stripped:
        raise PackArchiveError(f"Windows-ambiguous path component: {part!r}")
    stem = stripped.split(".", 1)[0].upper()
    if stem in WINDOWS_RESERVED:
        raise PackArchiveError(f"Windows-reserved path component: {part!r}")


def validate_archive_member_name(
    name: str,
    *,
    is_directory: bool,
    limits: PackSafetyLimits,
) -> tuple[str, tuple[str, ...]]:
    """Validate and canonicalize one ZIP member name without touching disk."""
    if not isinstance(name, str) or not name or "\x00" in name or "\\" in name:
        raise PackArchiveError(f"unsafe archive member name: {name!r}")
    if len(name) > limits.max_path_chars + 1:
        raise PackArchiveError("archive member path exceeds configured bound")
    if name.startswith("/") or DRIVE_QUALIFIED.match(name):
        raise PackArchiveError(f"absolute/drive-qualified archive member: {name!r}")

    if is_directory:
        if not name.endswith("/"):
            raise PackArchiveError(f"directory entry lacks canonical trailing slash: {name!r}")
        raw = name[:-1]
    else:
        if name.endswith("/"):
            raise PackArchiveError(f"file entry has directory-style path: {name!r}")
        raw = name

    if not raw or len(raw) > limits.max_path_chars:
        raise PackArchiveError("archive member path is empty or exceeds configured bound")
    if unicodedata.normalize("NFC", raw) != raw:
        raise PackArchiveError(f"archive member path is not NFC canonical: {name!r}")

    pure = PurePosixPath(raw)
    if pure.as_posix() != raw:
        raise PackArchiveError(f"non-canonical archive member path: {name!r}")
    parts = pure.parts
    if not parts or any(part in {"", ".", ".."} for part in parts):
        raise PackArchiveError(f"unsafe archive traversal path: {name!r}")
    for part in parts:
        _component_safe(part)

    if parts[0] not in ROOT_NAMESPACES:
        raise PackArchiveError(f"archive member outside metadata/targets namespaces: {name!r}")

    if parts[0] == "metadata" and not is_directory:
        if len(parts) != 2:
            raise PackArchiveError("Phase 5.5.2 v1 metadata files must be direct children of metadata/")
        metadata_name = parts[1]
        if metadata_name not in TOP_LEVEL_METADATA and not VERSIONED_ROOT_RE.fullmatch(metadata_name):
            raise PackArchiveError(f"unsupported v1 TUF metadata member: {metadata_name!r}")

    if parts[0] == "targets":
        if len(parts) < 2:
            if not is_directory:
                raise PackArchiveError("targets/ cannot itself be a file")
        elif parts[1] not in TARGET_NAMESPACES:
            raise PackArchiveError(f"target outside Atlas target namespaces: {name!r}")
        elif not is_directory and PurePosixPath(raw).suffix.casefold() in ACTIVE_EXTENSIONS:
            raise PackArchiveError(f"active-code extension is forbidden in content packs: {name!r}")

    return raw, parts


def _validate_entry_type(info: zipfile.ZipInfo) -> None:
    if info.flag_bits & 0x1:
        raise PackArchiveError(f"encrypted ZIP member is forbidden: {info.filename!r}")
    if info.compress_type not in ALLOWED_COMPRESSION:
        raise PackArchiveError(
            f"unsupported ZIP compression method {info.compress_type}: {info.filename!r}"
        )

    dos_attrs = info.external_attr & 0xFFFF
    if dos_attrs & WINDOWS_REPARSE_POINT:
        raise PackArchiveError(f"Windows reparse-point archive member is forbidden: {info.filename!r}")

    unix_mode = (info.external_attr >> 16) & 0xFFFF
    if unix_mode:
        file_type = stat.S_IFMT(unix_mode)
        allowed = {0, stat.S_IFREG, stat.S_IFDIR}
        if file_type not in allowed:
            raise PackArchiveError(f"non-regular archive member is forbidden: {info.filename!r}")
        if info.is_dir() and file_type not in {0, stat.S_IFDIR}:
            raise PackArchiveError(f"directory entry has non-directory mode: {info.filename!r}")
        if not info.is_dir() and file_type == stat.S_IFDIR:
            raise PackArchiveError(f"file entry has directory mode: {info.filename!r}")


def _preflight(
    archive: zipfile.ZipFile,
    limits: PackSafetyLimits,
) -> tuple[list[tuple[zipfile.ZipInfo, str, tuple[str, ...]]], int, int, int]:
    infos = archive.infolist()
    if len(infos) > limits.max_entries:
        raise PackArchiveError(f"archive has {len(infos)} entries; limit is {limits.max_entries}")

    seen: set[str] = set()
    validated: list[tuple[zipfile.ZipInfo, str, tuple[str, ...]]] = []
    total = 0
    files = 0
    directories = 0

    for info in infos:
        # CPython's ZIP parser may sanitize the raw member name (for example,
        # backslashes are converted to forward slashes on Windows and NUL can
        # truncate a name). Atlas validates the wire-level name, not a parser-
        # repaired representation. Any parser mutation is therefore fail-closed.
        original_name = getattr(info, "orig_filename", info.filename)
        if original_name != info.filename:
            raise PackArchiveError(
                f"ZIP parser sanitized a non-canonical archive member name: {original_name!r}"
            )

        _validate_entry_type(info)
        raw, parts = validate_archive_member_name(
            original_name, is_directory=info.is_dir(), limits=limits
        )
        collision_key = unicodedata.normalize("NFC", raw).casefold()
        if collision_key in seen:
            raise PackArchiveError(f"duplicate/case-colliding archive path: {info.filename!r}")
        seen.add(collision_key)

        if info.is_dir():
            if info.file_size != 0:
                raise PackArchiveError(f"directory entry has non-zero size: {info.filename!r}")
            directories += 1
        else:
            files += 1
            if info.file_size < 0 or info.file_size > limits.max_file_uncompressed_bytes:
                raise PackArchiveError(f"archive member exceeds per-file bound: {info.filename!r}")
            total += info.file_size
            if total > limits.max_total_uncompressed_bytes:
                raise PackArchiveError("archive exceeds aggregate uncompressed-size bound")
            if info.file_size:
                ratio = info.file_size / max(info.compress_size, 1)
                if ratio > limits.max_compression_ratio:
                    raise PackArchiveError(
                        f"archive member exceeds compression-ratio bound: {info.filename!r}"
                    )
        validated.append((info, raw, parts))

    return validated, files, directories, total


def _ensure_private_empty_directory(destination: Path) -> None:
    if destination.exists():
        if destination.is_symlink() or not destination.is_dir():
            raise PackArchiveError("staging destination is not a normal directory")
        if any(destination.iterdir()):
            raise PackArchiveError("staging destination must be empty")
    else:
        destination.mkdir(parents=True, mode=0o700)
    try:
        destination.chmod(0o700)
    except OSError:
        pass


def _mkdir_chain_without_links(root: Path, parts: tuple[str, ...]) -> Path:
    current = root
    for part in parts:
        current = current / part
        if current.exists() or current.is_symlink():
            if current.is_symlink() or not current.is_dir():
                raise PackArchiveError(f"staging path component is not a normal directory: {current}")
        else:
            current.mkdir(mode=0o700)
    return current


def safe_extract_atlaspack(
    archive_path: Path,
    destination: Path,
    *,
    limits: PackSafetyLimits | None = None,
) -> ExtractionReport:
    """Preflight every member, then manually stream a pack into private staging."""
    limits = limits or PackSafetyLimits()
    archive_path = Path(archive_path)
    destination = Path(destination)
    if not archive_path.is_file():
        raise PackArchiveError(f"atlaspack does not exist: {archive_path}")
    if archive_path.stat().st_size > limits.max_archive_bytes:
        raise PackArchiveError("atlaspack exceeds configured compressed-size bound")
    _ensure_private_empty_directory(destination)

    try:
        with zipfile.ZipFile(archive_path, "r") as archive:
            validated, files, directories, total = _preflight(archive, limits)
            for info, _raw, parts in validated:
                if info.is_dir():
                    _mkdir_chain_without_links(destination, parts)
                    continue

                parent = _mkdir_chain_without_links(destination, parts[:-1])
                target = parent / parts[-1]
                if target.exists() or target.is_symlink():
                    raise PackArchiveError(f"refusing to overwrite staging path: {target}")

                written = 0
                with archive.open(info, "r") as source, target.open("xb") as output:
                    while True:
                        chunk = source.read(COPY_CHUNK)
                        if not chunk:
                            break
                        written += len(chunk)
                        if written > info.file_size or written > limits.max_file_uncompressed_bytes:
                            raise PackArchiveError(
                                f"decompressed member exceeded declared/safety size: {info.filename!r}"
                            )
                        output.write(chunk)
                    output.flush()
                    os.fsync(output.fileno())
                if written != info.file_size:
                    raise PackArchiveError(
                        f"decompressed member size mismatch: {info.filename!r}: {written} != {info.file_size}"
                    )
                try:
                    target.chmod(0o600)
                except OSError:
                    pass
    except PackArchiveError:
        raise
    except (zipfile.BadZipFile, RuntimeError, OSError) as exc:
        raise PackArchiveError(f"atlaspack extraction failed: {exc}") from exc

    return ExtractionReport(
        profile_version=limits.profile_version,
        file_count=files,
        directory_count=directories,
        total_uncompressed_bytes=total,
    )

File: shared-core/go/internal/pack/archive.go
package pack

import (
	"archive/zip"
	"errors"
	"fmt"
	"io"
	"os"
	"path"
	"path/filepath"
	"regexp"
	"strings"

	"golang.org/x/text/unicode/norm"
)

const copyChunk = 1024 * 1024

var versionedRootRE = regexp.MustCompile(`^[1-9][0-9]*\.root\.json$`)

var topLevelMetadata = map[string]struct{}{
	"root.json": {}, "timestamp.json": {}, "snapshot.json": {}, "targets.json": {},
}

type SafetyLimits struct {
	ProfileVersion           string
	MaxArchiveBytes          int64
	MaxEntries               int
	MaxPathChars             int
	MaxFileUncompressedBytes uint64
	MaxTotalUncompressed     uint64
	MaxCompressionRatio      float64
}

func DefaultSafetyLimits() SafetyLimits {
	return SafetyLimits{
		ProfileVersion:           "1.0.0",
		MaxArchiveBytes:          2 * 1024 * 1024 * 1024,
		MaxEntries:               8192,
		MaxPathChars:             512,
		MaxFileUncompressedBytes: 2 * 1024 * 1024 * 1024,
		MaxTotalUncompressed:     4 * 1024 * 1024 * 1024,
		MaxCompressionRatio:      200.0,
	}
}

type ExtractionReport struct {
	ProfileVersion         string `json:"profile_version"`
	FileCount              int    `json:"file_count"`
	DirectoryCount         int    `json:"directory_count"`
	TotalUncompressedBytes uint64 `json:"total_uncompressed_bytes"`
}

type validatedEntry struct {
	file  *zip.File
	clean string
	parts []string
}

func validateArchiveMemberName(name string, isDirectory bool, limits SafetyLimits) (string, []string, error) {
	if name == "" || strings.ContainsRune(name, '\x00') || strings.Contains(name, "\\") {
		return "", nil, fmt.Errorf("unsafe archive member name: %q", name)
	}
	maxLen := limits.MaxPathChars
	if isDirectory {
		if !strings.HasSuffix(name, "/") {
			return "", nil, fmt.Errorf("directory entry lacks canonical trailing slash: %q", name)
		}
		name = strings.TrimSuffix(name, "/")
	} else if strings.HasSuffix(name, "/") {
		return "", nil, fmt.Errorf("file entry has directory-style path")
	}
	if name == "" || len(name) > maxLen || strings.HasPrefix(name, "/") || driveRE.MatchString(name) {
		return "", nil, fmt.Errorf("archive member path is empty/absolute/too long: %q", name)
	}
	if norm.NFC.String(name) != name || path.Clean(name) != name {
		return "", nil, fmt.Errorf("archive member path is not canonical NFC: %q", name)
	}
	parts := strings.Split(name, "/")
	for _, part := range parts {
		if err := validatePathComponent(part); err != nil {
			return "", nil, err
		}
	}
	if parts[0] != "metadata" && parts[0] != "targets" {
		return "", nil, fmt.Errorf("archive member outside metadata/targets namespaces: %q", name)
	}
	if parts[0] == "metadata" && !isDirectory {
		if len(parts) != 2 {
			return "", nil, fmt.Errorf("v1 metadata files must be direct children of metadata/")
		}
		if _, ok := topLevelMetadata[parts[1]]; !ok && !versionedRootRE.MatchString(parts[1]) {
			return "", nil, fmt.Errorf("unsupported v1 TUF metadata member: %q", parts[1])
		}
	}
	if parts[0] == "targets" {
		if len(parts) < 2 {
			if !isDirectory {
				return "", nil, fmt.Errorf("targets/ cannot itself be a file")
			}
		} else if parts[1] != "atlas" && parts[1] != "content" && parts[1] != "search" {
			return "", nil, fmt.Errorf("target outside Atlas namespaces: %q", name)
		} else if !isDirectory {
			if _, blocked := activeExtensions[strings.ToLower(path.Ext(name))]; blocked {
				return "", nil, fmt.Errorf("active-code extension is forbidden in content packs: %q", name)
			}
		}
	}
	return name, parts, nil
}

func validateZipEntryType(file *zip.File) error {
	if file.Flags&0x1 != 0 {
		return fmt.Errorf("encrypted ZIP member is forbidden: %q", file.Name)
	}
	if file.NonUTF8 {
		return fmt.Errorf("non-UTF8 ZIP member name is forbidden: %q", file.Name)
	}
	if file.Method != zip.Store && file.Method != zip.Deflate {
		return fmt.Errorf("unsupported ZIP compression method %d: %q", file.Method, file.Name)
	}
	// Preserve the Phase 5.5.2 cross-platform reparse-point guard.
	if file.ExternalAttrs&0x0400 != 0 {
		return fmt.Errorf("Windows reparse-point archive member is forbidden: %q", file.Name)
	}
	mode := file.Mode()
	if mode&os.ModeSymlink != 0 || mode&(os.ModeDevice|os.ModeNamedPipe|os.ModeSocket|os.ModeCharDevice) != 0 {
		return fmt.Errorf("non-regular archive member is forbidden: %q", file.Name)
	}
	return nil
}

func preflightArchive(reader *zip.ReadCloser, limits SafetyLimits) ([]validatedEntry, ExtractionReport, error) {
	if len(reader.File) > limits.MaxEntries {
		return nil, ExtractionReport{}, fmt.Errorf("archive has %d entries; limit is %d", len(reader.File), limits.MaxEntries)
	}
	report := ExtractionReport{ProfileVersion: limits.ProfileVersion}
	seen := map[string]struct{}{}
	validated := make([]validatedEntry, 0, len(reader.File))
	for _, file := range reader.File {
		if err := validateZipEntryType(file); err != nil {
			return nil, ExtractionReport{}, err
		}
		isDir := file.FileInfo().IsDir()
		clean, parts, err := validateArchiveMemberName(file.Name, isDir, limits)
		if err != nil {
			return nil, ExtractionReport{}, err
		}
		collisionKey := strings.ToLower(norm.NFC.String(clean))
		if _, exists := seen[collisionKey]; exists {
			return nil, ExtractionReport{}, fmt.Errorf("duplicate/case-colliding archive path: %q", clean)
		}
		seen[collisionKey] = struct{}{}
		if isDir {
			if file.UncompressedSize64 != 0 {
				return nil, ExtractionReport{}, fmt.Errorf("directory entry has non-zero size: %q", clean)
			}
			report.DirectoryCount++
		} else {
			report.FileCount++
			if file.UncompressedSize64 > limits.MaxFileUncompressedBytes {
				return nil, ExtractionReport{}, fmt.Errorf("archive member exceeds per-file bound: %q", clean)
			}
			if ^uint64(0)-report.TotalUncompressedBytes < file.UncompressedSize64 {
				return nil, ExtractionReport{}, fmt.Errorf("archive size accounting overflow")
			}
			report.TotalUncompressedBytes += file.UncompressedSize64
			if report.TotalUncompressedBytes > limits.MaxTotalUncompressed {
				return nil, ExtractionReport{}, fmt.Errorf("archive exceeds aggregate uncompressed-size bound")
			}
			if file.UncompressedSize64 > 0 {
				compressed := file.CompressedSize64
				if compressed == 0 {
					compressed = 1
				}
				ratio := float64(file.UncompressedSize64) / float64(compressed)
				if ratio > limits.MaxCompressionRatio {
					return nil, ExtractionReport{}, fmt.Errorf("archive member exceeds compression-ratio bound: %q", clean)
				}
			}
		}
		validated = append(validated, validatedEntry{file: file, clean: clean, parts: parts})
	}
	return validated, report, nil
}

func ensurePrivateEmptyDirectory(destination string) error {
	info, err := os.Lstat(destination)
	if err == nil {
		if !info.IsDir() || info.Mode()&os.ModeSymlink != 0 {
			return fmt.Errorf("staging destination is not a normal directory")
		}
		entries, err := os.ReadDir(destination)
		if err != nil {
			return err
		}
		if len(entries) != 0 {
			return fmt.Errorf("staging destination must be empty")
		}
		return os.Chmod(destination, 0o700)
	}
	if !errors.Is(err, os.ErrNotExist) {
		return err
	}
	return os.MkdirAll(destination, 0o700)
}

func mkdirChainWithoutLinks(root string, parts []string) (string, error) {
	current := root
	for _, part := range parts {
		current = filepath.Join(current, part)
		info, err := os.Lstat(current)
		if err == nil {
			if !info.IsDir() || info.Mode()&os.ModeSymlink != 0 {
				return "", fmt.Errorf("staging path component is not a normal directory: %s", current)
			}
			continue
		}
		if !errors.Is(err, os.ErrNotExist) {
			return "", err
		}
		if err := os.Mkdir(current, 0o700); err != nil {
			return "", err
		}
	}
	return current, nil
}

func SafeExtractAtlaspack(archivePath, destination string, limits SafetyLimits) (ExtractionReport, error) {
	info, err := os.Stat(archivePath)
	if err != nil || !info.Mode().IsRegular() {
		return ExtractionReport{}, fmt.Errorf("atlaspack does not exist or is not regular")
	}
	if info.Size() > limits.MaxArchiveBytes {
		return ExtractionReport{}, fmt.Errorf("atlaspack exceeds configured compressed-size bound")
	}
	if err := ensurePrivateEmptyDirectory(destination); err != nil {
		return ExtractionReport{}, err
	}
	reader, err := zip.OpenReader(archivePath)
	if err != nil {
		return ExtractionReport{}, fmt.Errorf("atlaspack open failed: %w", err)
	}
	defer reader.Close()
	validated, report, err := preflightArchive(reader, limits)
	if err != nil {
		return ExtractionReport{}, err
	}
	for _, entry := range validated {
		if entry.file.FileInfo().IsDir() {
			if _, err := mkdirChainWithoutLinks(destination, entry.parts); err != nil {
				return ExtractionReport{}, err
			}
			continue
		}
		parent, err := mkdirChainWithoutLinks(destination, entry.parts[:len(entry.parts)-1])
		if err != nil {
			return ExtractionReport{}, err
		}
		target := filepath.Join(parent, entry.parts[len(entry.parts)-1])
		if _, err := os.Lstat(target); err == nil {
			return ExtractionReport{}, fmt.Errorf("refusing to overwrite staging path: %s", target)
		} else if !errors.Is(err, os.ErrNotExist) {
			return ExtractionReport{}, err
		}
		source, err := entry.file.Open()
		if err != nil {
			return ExtractionReport{}, err
		}
		output, err := os.OpenFile(target, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
		if err != nil {
			source.Close()
			return ExtractionReport{}, err
		}
		written, copyErr := io.CopyBuffer(output, io.LimitReader(source, int64(entry.file.UncompressedSize64)+1), make([]byte, copyChunk))
		syncErr := output.Sync()
		closeOutErr := output.Close()
		closeSourceErr := source.Close()
		if copyErr != nil {
			return ExtractionReport{}, copyErr
		}
		if written != int64(entry.file.UncompressedSize64) {
			return ExtractionReport{}, fmt.Errorf("decompressed member size mismatch: %q", entry.clean)
		}
		if syncErr != nil {
			return ExtractionReport{}, syncErr
		}
		if closeOutErr != nil {
			return ExtractionReport{}, closeOutErr
		}
		if closeSourceErr != nil {
			return ExtractionReport{}, closeSourceErr
		}
	}
	return report, nil
}
